Document Sec-CH-Prefers-Color-Scheme Client Hint header (#22266)
mdn/content: files/en-us/web/http/headers/sec-ch-prefers-color-scheme/index.md

---
title: Sec-CH-Prefers-Color-Scheme
slug: Web/HTTP/Headers/Sec-CH-Prefers-Color-Scheme
tags:
  - Sec-CH-Prefers-Color-Scheme
  - Client hint
  - HTTP
  - HTTP Header
  - prefers-color-scheme
  - Reference
  - Request header
  - Experimental
browser-compat: http.headers.Sec-CH-Prefers-Color-Scheme
---

{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}

> The **`Sec-CH-Prefers-Color-Scheme`** [user agent client hint](/en-US/docs/Web/HTTP/Client_hints#user-agent_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (e.g. light or dark mode) or a user agent setting.

Reviewer: ```suggestion
The **`Sec-CH-Prefers-Color-Scheme`** [user preference media feature client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (for example, light or dark mode) or a user agent setting.
```

      <td>yes</td>
    </tr>
  </tbody>
</table>

The **`Sec-CH-Prefers-Color-Scheme`** header expesses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particular color scheme. Value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value useed in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.

`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theiry be used for user fingerprinting.

## Syntax

```http
Sec-CH-Prefers-Color-Scheme: <preference>
```

### Directives

- `<preference>`

>   - : The user agent's preference for dark or light content. This is often taken from the underlying operating system's setting. The value of this directive can be either `light` or `dark`. A string indicating the user preference, either `"light"` or `"dark"`.

Reviewer: I have a preference for stating the values and then the source.

```suggestion
  - : A string indicating the user agent's preference for dark or light content: `"light"` or `"dark"`.
    The value may originate from a corresponding setting in the underlying operating system.
```

## Examples

The client makes an initial request to the server:

```http
GET / HTTP/1.1
Host: example.com
```

> The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Color-Scheme`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Prefers-Color-Scheme` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints).

Reviewer: FYI I think it makes sense to use Critical-CH in this example (as you have done) because the user has very clearly inidicated this preference, and you'll get a jarring disconnect if you navigate to another page on the site.

```http
HTTP/1.1 200 OK
Content-Type: text/html
Accept-CH: Sec-CH-Prefers-Color-Scheme
Vary: Sec-CH-Prefers-Color-Scheme
Critical-CH: Sec-CH-Prefers-Color-Scheme
```

> **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.

The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Prefers-Color-Scheme` that it has a user preference for dark content:

```http
GET / HTTP/1.1
Host: example.com
Sec-CH-Prefers-Color-Scheme: "dark"
```

> The client will include the header in subsequent requests in the current session unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.

Reviewer: ```suggestion
The client will include the header in subsequent requests in the current session, unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.
```

---
title: Sec-CH-Prefers-Color-Scheme
slug: Web/HTTP/Headers/Sec-CH-Prefers-Color-Scheme
tags:
  - Sec-CH-Prefers-Color-Scheme
  - Client hint
  - HTTP
  - HTTP Header
  - prefers-color-scheme
  - Reference
  - Request header
  - Experimental
browser-compat: http.headers.Sec-CH-Prefers-Color-Scheme
---

{{HTTPSidebar}}{{SeeCompatTable}}{{SecureContext_Header}}

The **`Sec-CH-Prefers-Color-Scheme`** [user preference media feature client hint](/en-US/docs/Web/HTTP/Client_hints#user_preference_media_features_client_hints) request header provides the user's preference for light or dark color themes. A user indicates their preference through an operating system setting (for example, light or dark mode) or a user agent setting.

> If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Color-Scheme`, the client can then respond with this header to indicate the user's preference for a specific color scheme. The server can send the client appropriately adapted content, for example, images or CSS, to display light text on dark background on subsequent rendered content.

Reviewer 2: The last sentence here is a bit awkward to read because of all the commas. How about "The server can send the client appropriately adapted content including images or CSS to display a light or dark mode for subsequent rendered content." ?

Reviewer: Yes, that is more readable.

```suggestion
If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Color-Scheme`, the client can then respond with this header to indicate the user's preference for a specific color scheme. The server can send the client appropriately adapted content including images or CSS to display a light or dark mode for subsequent rendered content.
```

This header is modeled on the {{cssxref("@media/prefers-color-scheme", "prefers-color-scheme")}} media query.

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>
        {{Glossary("Request header")}},
        <a href="/en-US/docs/Web/HTTP/Client_hints">Client hint</a>
      </td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>yes</td>
    </tr>
  </tbody>
</table>

> The **`Sec-CH-Prefers-Color-Scheme`** header expesses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particular color scheme. Value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value useed in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.

Reviewer 2: * "If server inlines CSS like that, it might want to to include {{HTTPHeader("Vary")}} header" -> "If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header"
* "Value reported by" -> "The value reported by"
* "used" -> "used"

Reviewer: Integrates Chris suggestions. I tend to split these up paragraph into per-line sentences - but that is my taste - makes reviewing easier.

```suggestion
The **`Sec-CH-Prefers-Color-Scheme`** header expresses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particular color scheme.
The value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value used in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.
```

> `Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theiry be used for user fingerprinting.

Reviewer 2: "theiry" -> "theory"

Reviewer: @chrisdavidmills FWIW I just fix the obvious typos and merge them. Saves a little round tripping.
```suggestion
`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theory be used for user fingerprinting.
```

Content-Type: text/html
Accept-CH: Sec-CH-Prefers-Color-Scheme
Vary: Sec-CH-Prefers-Color-Scheme
Critical-CH: Sec-CH-Prefers-Color-Scheme
```

> **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.

The client automatically retries the request (due to `Critical-CH` being specified above), telling the server via `Sec-CH-Prefers-Color-Scheme` that it has a user preference for dark content:

```http
GET / HTTP/1.1
Host: example.com
Sec-CH-Prefers-Color-Scheme: "dark"
```

The client will include the header in subsequent requests in the current session, unless the `Accept-CH` changes in responses to indicate that it is no longer supported by the server.

## Specifications

{{Specifications}}

## Browser compatibility

{{Compat}}

## See also

- [Client hints](/en-US/docs/Web/HTTP/Client_hints)

> - `prefers-color-scheme` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-color-scheme)

Reviewer 2: Missing an opening square bracket here.

Reviewer: FWIW fix as suggestion and self merge if there is "no question" it is just a bug. This is the "mike smith" review method :-)

```suggestion
- [`prefers-color-scheme` CSS Media Query](/en-US/docs/Web/CSS/@media/prefers-color-scheme)
```

If a server signals to a client via the {{httpheader("Accept-CH")}} header that it accepts `Sec-CH-Prefers-Color-Scheme`, the client can then respond with this header to indicate the user's preference for a specific color scheme. The server can send the client appropriately adapted content including images or CSS to display a light or dark mode for subsequent rendered content.

This header is modeled on the {{cssxref("@media/prefers-color-scheme", "prefers-color-scheme")}} media query.

<table class="properties">
  <tbody>
    <tr>
      <th scope="row">Header type</th>
      <td>
        {{Glossary("Request header")}},
        <a href="/en-US/docs/Web/HTTP/Client_hints">Client hint</a>
      </td>
    </tr>
    <tr>
      <th scope="row">{{Glossary("Forbidden header name")}}</th>
      <td>yes</td>
    </tr>
  </tbody>
</table>

> The **`Sec-CH-Prefers-Color-Scheme`** header expresses the user's media preference to see content in dark or light color schemes, allowing sites to obtain it at request time to inline the right CSS for performance reasons. If the server inlines CSS, it might want to include the {{HTTPHeader("Vary")}} header with `Sec-CH-Prefers-Color-Scheme` to specify that the response is tailored for a particular color scheme.

Reviewer: First sentence feels repetitive - i.e we already said this. Also seems a bit odd to have some content below the table and some above it. Usually if there is too much to say above the table then there is a separate "Description" section for the detail, and the information above the table is kept very succint.

Reviewer 2: In my most recent commit, I updated this section to make it less repetitive and be more about using it over the media query version if performance is a critical consideration in this context, and what other headers are needed alongside it. I also placed it under a "Usage notes" heading.

The value reported by `Sec-CH-Prefers-Color-Scheme` should be equivalent to the same value used in CSS Media Queries [`prefers-color-scheme`](/en-US/docs/Web/CSS/@media/prefers-color-scheme) and the JavaScript API {{domxref("Window.matchMedia()")}}.

> `Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. The user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy since the reported user preference could in theory be used for user fingerprinting.

Reviewer: ```suggestion
`Sec-CH-Prefers-Color-Scheme` is a high entropy hint so the site needs to opt into receiving it by sending an appropriate response header {{HTTPHeader("Accept-CH")}}. A user agent may intentionally omit the `Sec-CH-Prefers-Color-Scheme` header to preserve user privacy, since the user's preference could in theory be used for fingerprinting.
```

## Syntax

```http
Sec-CH-Prefers-Color-Scheme: <preference>
```

### Directives

- `<preference>`

  - : A string indicating the user agent's preference for dark or light content: `"light"` or `"dark"`.
    The value may originate from a corresponding setting in the underlying operating system.

## Examples

The client makes an initial request to the server:

```http
GET / HTTP/1.1
Host: example.com
```

The server responds, telling the client via {{httpheader("Accept-CH")}} that it accepts `Sec-CH-Prefers-Color-Scheme`. In this example {{httpheader("Critical-CH")}} is also used, indicating that `Sec-CH-Prefers-Color-Scheme` is considered a [critical client hint](/en-US/docs/Web/HTTP/Client_hints#critical_client_hints).

```http
HTTP/1.1 200 OK
Content-Type: text/html
Accept-CH: Sec-CH-Prefers-Color-Scheme
Vary: Sec-CH-Prefers-Color-Scheme
Critical-CH: Sec-CH-Prefers-Color-Scheme
```

> > **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate to the browser that the served content will differ based on this header value, even if the URL stays the same, so the browser shouldn't just use an existing cached response and instead should cache this response separately. Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.

Reviewer: ```suggestion
> **Note:** We've also specified `Sec-CH-Prefers-Color-Scheme` in the {{httpheader("Vary")}} header to indicate that responses should be separately cached based on the value of this header (even if the URL stays the same). 
> Each header listed in the `Critical-CH` header should also be present in the `Accept-CH` and `Vary` headers.
```

> ---

Reviewer: @chrisdavidmills A good starting point for this doc?
